Adds documentation for Sway attributes. (#4203)
FuelLabs/sway: docs/book/src/reference/attributes.md

# Attributes

> The Sway compiler supports a list of attributes that perform various operations that are useful for building, testing and documenting libraries. Below is a list of all available attributes:

Reviewer: ```suggestion
The Sway compiler supports a list of attributes that perform various operations that are useful for building, testing and documenting Sway programs. Below is a list of all available attributes:
```

## Allow

The `#[allow(dead_code)]` attribute overrides the check for dead code so that violations will go unreported.

## Doc

The `#[doc(..)]` attribute specifies documentation.

Line doc comments beginning with exactly three slashes `///`, are interpreted as a special syntax for doc attributes. That is, they are equivalent to writing `#[doc("...")]` around the body of the comment, i.e., `/// Foo` turns into `#[doc("Foo")]`

> Line comments beginning with `//!` are doc comments that apply to the parent of the comment, rather than the item that follows. That is, they are equivalent to writing `#![doc("...")]` around the body of the comment. `//!` comments are usually used to document modules that occupy a source file.

Reviewer: cc @eureka-cpu - is this an accurate description of how things work in Sway?

Reviewer 2: So it's not that they are usually used, but they can _only_ be used for documenting modules.

Author: Removed the part about `//!` and `#![doc("...")]`.

Reviewer: Should we completely remove them completely or simply limit their scope to modules?

Reviewer 2: I think we should include it as @mohammadfawaz suggested

Reviewer 2: If you do, it would also close #4114

Author: Added again the `//!` part, with scope limited to modules and mentioned that it should be used at the beginning of Sway files.

> Documentation can be generated from doc attributes using `forc doc`.

Reviewer 2: This may be a good place to have linked to the documentation about `forc doc` since this command will only generate the documentation and not open it.

## Inline

The inline attribute suggests that a copy of the attributed function should be placed in the caller, rather than generating code to call the function where it is defined.

> **Note**: The Sway compiler automatically inlines functions based on internal heuristics. Incorrectly inlining functions can make the program slower, so this attribute should be used with care.

> The `#[inline(never)]` attribute *suggests* that an inline expansion should never be
>   performed.

Reviewer: ```suggestion
The `#[inline(never)]` attribute *suggests* that an inline expansion should never be performed.
```

> The `#[inline(always)]` attribute *suggests* that an inline expansion should always be
>   performed.

Reviewer: ```suggestion
The `#[inline(always)]` attribute *suggests* that an inline expansion should always be performed.
```

> **Note**: `#[inline(..)]` in every form is a hint, with no *requirements*
 on the language to place a copy of the attributed function in the caller.

## Payable

> The lack of `#[payable]` implies the method is non-payable. When calling an ABI method that is non-payable, the compiler emits an error if the amount of coins to forward is not guaranteed to be zero.

Reviewer: ```suggestion
The lack of `#[payable]` implies the method is non-payable. When calling an ABI method that is non-payable, the compiler emits an error if the amount of coins forwarded with the call is not guaranteed to be zero. Note that this is strictly a compile-time check and does not incur any runtime cost.
```

## Storage

In Sway, functions are pure by default but can be opted into impurity via the `storage` function attribute. The `storage` attribute may take `read` and/or `write` arguments indicating which type of access the function requires.

The `#[storage(read)]` attribute indicates that a function requires read access to the storage.

The `#[storage(write)]` attribute indicates that a function requires write access to the storage.

More details in [Purity](../blockchain-development/purity.md).

## Test

> The `#[test]` attribute marks a function to be executed as a test.

Reviewer: We should probably mention `#[test(should_revert)]` here as well.
